### PR TITLE
[ROCm] Initial port of CUDAExtensions to HIP

### PR DESCRIPTION
**Pull request by iotamudelta**
After discussions with @goldsborough and @ezyang port the CUDAExtensions feature in-place to HIP. This entails introducing a HIP_COMP boolean that, identical to the build process, checks for the ROCm install directory to be present.

Add the default ROCm location and HIP flags in line with the regular PyTorch build.

Add .hip as a recognized file extension as that is what is used internally.

### CODE REVIEW

File: torch/utils/cpp_extension.py [hunk 1]
```diff
@@ -85,8 +85,9 @@ def _find_cuda_home():
 
                               !! WARNING !!
 '''
-CUDA_HOME = _find_cuda_home()
-CUDNN_HOME = os.environ.get('CUDNN_HOME') or os.environ.get('CUDNN_PATH')
+HIP_COMP = os.path.exists('/opt/rocm')
```
Comment (ezyang) [inline torch/utils/cpp_extension.py]: This feels wrong. The presence of `/opt/rocm` folder doesn't mean that I didn't want CUDA. You should be querying whether or not the PyTorch build was a CUDA or ROCM build and act accordingly (or, in a pinch, toggle this by environment variable.)

**Review by ezyang — CHANGES_REQUESTED** [review]
Everything else looks reasonable, but not `HIP_COMP` detection.

**Comment by ezyang** [comment]
Also, I didn't see anything that would have turned on these tests in ROCm CI. If you care about making sure this keeps working, we'd better enable the relevant tests.

**Comment by bddppq** [comment]
torchvision needs to be "hipified" as well. [Here](https://github.com/pytorch/vision/blob/b5db97b/setup.py#L104) it thinks with `CUDA_HOME` set, the build is cuda (and adds `-DWITH_CUDA` to the build command), and then triggers the cuda include [here](https://github.com/pytorch/vision/blob/b3cb86f/torchvision/csrc/vision.cpp#L5-L7).
``` 
00:33:41   x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DWITH_CUDA -I/tmp/pip-req-build-MqIE65/torchvision/csrc -I/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include -I/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/torch/csrc/api/include -I/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/TH -I/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/THC -I/opt/rocm/include -I/opt/rocm/miopen/include -I/usr/include/python2.7 -c /tmp/pip-req-build-MqIE65/torchvision/csrc/vision.cpp -o build/temp.linux-x86_64-2.7/tmp/pip-req-build-MqIE65/torchvision/csrc/vision.o -O0 -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_C -D_GLIBCXX_USE_CXX11_ABI=1 -std=c++11
00:33:41   cc1plus: warning: command line option â-Wstrict-prototypesâ is valid for C/ObjC but not for C++
00:33:41   In file included from /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/c10/cuda/impl/CUDAGuardImpl.h:6:0,
00:33:41                    from /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/c10/cuda/CUDAGuard.h:3,
00:33:41                    from /tmp/pip-req-build-MqIE65/torchvision/csrc/cuda/vision.h:2,
00:33:41                    from /tmp/pip-req-build-MqIE65/torchvision/csrc/ROIAlign.h:6,
00:33:41                    from /tmp/pip-req-build-MqIE65/torchvision/csrc/vision.cpp:1:
00:33:41   /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/include/c10/cuda/CUDAException.h:5:18: fatal error: cuda.h: No such file or directory
00:33:41   compilation terminated.
00:33:41   error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

**Review by bddppq — COMMENTED** [review]
Need to add hip build to torchvision since torchvision builds its c++ sources as cpp extension.

**Comment by ezyang** [comment]
Talk of HIPifying torchvision makes me very nervous. Do you *absolutely* need the CUDA kernels that live in torchvision? Can we perhaps just add a "no C++" build option to torchvision?

**Comment by iotamudelta** [comment]
@ezyang going forward, we'd like to support the detection networks in torchvision on the GPUs - wouldn't we need the CUDA/HIP kernels for that?

File: torch/utils/cpp_extension.py (hunk 1 above)
Comment (iotamudelta) [reply torch/utils/cpp_extension.py]: I see your point, I'd love for a reliable way to detect HIP compilation, I am not sure there is one currently. So maybe environment variable...

TBF, the build scripts rely on detecting `/opt/rocm` to switch the HIP build on currently.

**Comment by iotamudelta** [comment]
@ezyang which tests do you want to turn on?

File: torch/utils/cpp_extension.py (hunk 1 above)
Comment (t-vi) [reply torch/utils/cpp_extension.py]: So this should be easy by testing `version.hip is not None` now...

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -254,13 +262,16 @@ def unix_wrap_compile(obj, src, ext, cc_args, extra_postargs, pp_opts):
...
                     if not isinstance(nvcc, list):
                         nvcc = [nvcc]
                     self.compiler.set_executable('compiler_so', nvcc)
                     if isinstance(cflags, dict):
                         cflags = cflags['nvcc']
-                    cflags = COMMON_NVCC_FLAGS + ['--compiler-options', "'-fPIC'"] + cflags
+                    if HIP_COMP:
+                       cflags = COMMON_HIPCC_FLAGS + cflags
+                    else:
+                       cflags = COMMON_NVCC_FLAGS + ['--compiler-options', "'-fPIC'"] + cflags
```
Comment (t-vi) [inline torch/utils/cpp_extension.py]: There is a new merge conflict here from master adding `+ _get_cuda_arch_flags(cflags)`.

**Comment by iotamudelta** [comment]
Superseeded by #32669 - thanks for the review here, it was important to get it done properly!

**Pull request closed without merging**